src/main/java/jcifs/smb1/netbios/NameServiceClient.java

```diff
@@ -0,0 +1,444 @@
+            n = RETRY_COUNT;
+        } else {
+            request.isBroadcast = false;
+            n = 1;
+        }
+
+        do {
+            try {
+                send( request, response, RETRY_TIMEOUT );
+            } catch( IOException ioe ) {
+                if( log.level > 1 )
+                    ioe.printStackTrace( log );
+                throw new UnknownHostException( name.name );
+            }
+
+            if( response.received && response.resultCode == 0 ) {
+                return response.addrEntry;
+            }
+        } while( --n > 0 && request.isBroadcast );
+
+        throw new UnknownHostException( name.name );
+    }
+    NbtAddress getByName( Name name, InetAddress addr )
+                                            throws UnknownHostException {
+        int n;
+        NameQueryRequest request = new NameQueryRequest( name );
+        NameQueryResponse response = new NameQueryResponse();
+
+        if( addr != null ) { /* UniAddress calls always use this
+                              * because it specifies addr
+                              */
+            request.addr = addr; /* if addr ends with 255 flag it bcast */
+            request.isBroadcast = (addr.getAddress()[3] == (byte)0xFF);
+
+            n = RETRY_COUNT;
+            do {
+                try {
+                    send( request, response, RETRY_TIMEOUT );
+                } catch( IOException ioe ) {
+                    if( log.level > 1 )
+                        ioe.printStackTrace( log );
+                    throw new UnknownHostException( name.name );
+                }
+
+                if( response.received && response.resultCode == 0 ) {
+                    int last = response.addrEntry.length - 1;
+                    response.addrEntry[last].hostName.srcHashCode = addr.hashCode();
+                    return response.addrEntry[last];
+                }
+            } while( --n > 0 && request.isBroadcast );
+
+            throw new UnknownHostException( name.name );
+        }
+
+        /* If a target address to query was not specified explicitly
+         * with the addr parameter we fall into this resolveOrder routine.
+         */
+
+        for( int i = 0; i < resolveOrder.length; i++ ) {
+            try {
+                switch( resolveOrder[i] ) {
+                    case RESOLVER_LMHOSTS:
+                        NbtAddress ans = Lmhosts.getByName( name );
+                        if( ans != null ) {
+                            ans.hostName.srcHashCode = 0; // just has to be different
+                                                          // from other methods
+                            return ans;
+                        }
+                        break;
+                    case RESOLVER_WINS:
+                    case RESOLVER_BCAST:
+                        if( resolveOrder[i] == RESOLVER_WINS &&
+                                name.name != NbtAddress.MASTER_BROWSER_NAME &&
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.